rust/src/ima.rs

```diff
@@ -0,0 +1,234 @@
+        map.insert((*k).into(), (*v).into());
+
+        let map: Vec<_> = map.into_iter().collect();
+        variant_utils::new_variant_a_ayay(&map)
+    }
+
+    fn map_file(&mut self, checksum: &str) -> Result<Rc<str>> {
+        if let Some(r) = self.rewritten_files.get(checksum) {
+            return Ok(Rc::clone(r));
+        }
+        let cancellable = gio::NONE_CANCELLABLE;
+        let (instream, meta, xattrs) = self.repo.load_file(checksum, cancellable)?;
+        let instream =
+            instream.unwrap_or_else(|| gio::MemoryInputStream::new().upcast::<gio::InputStream>());
+        let meta = meta.unwrap();
+        let xattrs = xattrs.unwrap();
+
+        let xattrs = self.map_xattrs(&xattrs);
+
+        let (ostream, size) =
+            ostree::raw_file_to_content_stream(&instream, &meta, Some(&xattrs), cancellable)?;
+        let new_checksum = self
+            .repo
+            .write_content(None, &ostream, size, cancellable)?
+            .to_hex();
+
+        let r: Rc<str> = new_checksum.into();
+        self.rewritten_files
+            .insert(checksum.to_string(), Rc::clone(&r));
+        Ok(r)
+    }
+
+    /// Write a dirtree object.
+    fn map_dirtree(&mut self, checksum: &str) -> Result<String> {
+        let src = &self
+            .repo
+            .load_variant(ostree::ObjectType::DirTree, checksum)?;
+        let src = src.get_data_as_bytes();
+        let src = src.try_as_aligned()?;
+        let src = gv!("(a(say)a(sayay))").cast(src);
+        let (files, dirs) = src.to_tuple();
+
+        // A reusable buffer to avoid heap allocating these
+        let mut hexbuf = [0u8; 64];
+
+        let new_files_builder =
+            unsafe { glib_sys::g_variant_builder_new(b"a(say)\0".as_ptr() as *const _) };
+        for file in files {
+            let (name, csum) = file.to_tuple();
+            let name = name.to_str();
+            hex::encode_to_slice(csum, &mut hexbuf)?;
+            let checksum = std::str::from_utf8(&hexbuf)?;
+            let mapped = self.map_file(checksum)?;
+            let mapped = hex::decode(&*mapped)?;
+            unsafe {
+                // Unwrap safety: The name won't have NULs
+                let name = CString::new(name).unwrap();
+                let mapped_checksum_v = variant_utils::new_variant_bytearray(&mapped);
+                let name_p = name.as_ptr();
+                glib_sys::g_variant_builder_add(
+                    new_files_builder,
+                    b"(s@ay)\0".as_ptr() as *const _,
+                    name_p,
+                    mapped_checksum_v.to_glib_none().0,
+                );
+            }
+        }
+        let new_files: glib::Variant = unsafe {
+            let v = glib_sys::g_variant_builder_end(new_files_builder);
+            glib_sys::g_variant_ref_sink(v);
+            from_glib_full(v)
+        };
+
+        let new_dirs_builder =
+            unsafe { glib_sys::g_variant_builder_new(b"a(sayay)\0".as_ptr() as *const _) };
```

Comment:
This leaks on error paths.  We could probably add a thin wrapper struct which just proxies the unsafe C `add()` API.